### PR TITLE
MINOR: Add section on listener configuration (including kraft) to security docs

### PR DESCRIPTION
**Pull request by hachikuji**
This patch adds a section in security.html about listener configuration. This includes the basics of how to define the security mapping of each listener as well as the configurations to control inter-cluster traffic.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: docs/security.html [hunk 1]
```diff
@@ -36,7 +36,101 @@ <h3 class="anchor-heading"><a id="security_overview" class="anchor-link"></a><a
...
+      to the brokers. This is defined by <code>control.plane.listener.name</code>. The active controller
+      will use this listener when it needs to push metadata updates to the brokers in the cluster.
+      The benefit of using a control plane listener is that it uses a separate processing thread,
+      which makes it less likely for application traffic to impede timely propagation of metadata changes
+      (such as partition leader and ISR updates).</p>
+    
+    <p>In a KRaft cluster, the listener defined by <code>inter.broker.listener.name</code> is used
+      exclusively for requests sent to nodes which have the "broker" role enabled in <code>process.roles</code>.
+      Unlike with legacy clusters, brokers connect to controllers (i.e. nodes with the "controller" role enabled)
+      in order to receive metadata updates; not the other way around. Hence the configuration
```
Comment (showuon) [inline docs/security.html]: in order to -> is order to ?

File: docs/security.html [hunk 2]
```diff
@@ -36,7 +36,101 @@ <h3 class="anchor-heading"><a id="security_overview" class="anchor-link"></a><a
...
+      by the security protocol defined by <code>security.inter.broker.protocol</code>, which
+      defaults to <code>PLAINTEXT</code>.</p>
+    
+    <p>For legacy clusters which rely on Zookeeper to store cluster metadata, it is possible to
+      declare a separate listener to be used for metadata propagation from the active controller
+      to the brokers. This is defined by <code>control.plane.listener.name</code>. The active controller
+      will use this listener when it needs to push metadata updates to the brokers in the cluster.
+      The benefit of using a control plane listener is that it uses a separate processing thread,
+      which makes it less likely for application traffic to impede timely propagation of metadata changes
+      (such as partition leader and ISR updates).</p>
```
Comment (showuon) [inline docs/security.html]: Should we mention the `control.plane.listener.name` default value is null and there will be no dedicated endpoints for controller connections?

File: docs/security.html [hunk 3]
```diff
@@ -36,7 +36,101 @@ <h3 class="anchor-heading"><a id="security_overview" class="anchor-link"></a><a
...
+    
+    <p>In a KRaft cluster, the listener defined by <code>inter.broker.listener.name</code> is used
+      exclusively for requests sent to nodes which have the "broker" role enabled in <code>process.roles</code>.
+      Unlike with legacy clusters, brokers connect to controllers (i.e. nodes with the "controller" role enabled)
+      in order to receive metadata updates; not the other way around. Hence the configuration
+      <code>control.plane.listener.name</code> cannot be used in KRaft clusters.</p>
+    
+    <p>Instead, nodes with the "controller" role must define <code>controller.listener.names</code>.
+      Any node running with both the "broker" and "controller" roles must have use separate listeners
```
Comment (showuon) [inline docs/security.html]: must have use -> must use?

File: docs/security.html [hunk 4]
```diff
@@ -36,7 +36,101 @@ <h3 class="anchor-heading"><a id="security_overview" class="anchor-link"></a><a
...
+      for the broker an controller. For example, a minimal configuration for a combined node
```
Comment (showuon) [inline docs/security.html]: for the broker an[d] controller

**Review by showuon — COMMENTED** [review]
Thanks for adding the listener configuration section! It looks great and clear! Left some comments.

File: docs/security.html (hunk 1 above)
Comment (hachikuji) [reply docs/security.html]: Hmm, "in order to" seems right? Let me rephrase without the parenthetical to try and make it clearer.

File: docs/security.html (hunk 2 above)
Comment (hachikuji) [reply docs/security.html]: Good idea.

File: docs/security.html [hunk 5]
```diff
@@ -36,7 +36,104 @@ <h3 class="anchor-heading"><a id="security_overview" class="anchor-link"></a><a
 
     The guides below explain how to configure and use the security features in both clients and brokers.
 
-    <h3 class="anchor-heading"><a id="security_ssl" class="anchor-link"></a><a href="#security_ssl">7.2 Encryption and Authentication using SSL</a></h3>
+    <h3 class="anchor-heading"><a id="listener_configuration" class="anchor-link"></a><a href="#listener_configuration">7.2 Listener Configuration</a></h3>
+
+    <p>In order to secure a Kafka cluster, it is necessary to secure the channels that are used to
+      communicate with the servers. Each node must define the set of listeners that are used to
```
Comment (jsancio) [inline docs/security.html]:
```suggestion
      communicate with the servers. Each server must define the set of listeners that are used to
```

File: docs/security.html [hunk 6]
```diff
@@ -36,7 +36,104 @@ <h3 class="anchor-heading"><a id="security_overview" class="anchor-link"></a><a
...
+      receive requests from clients as well as other servers. Each listener may be configured
+      to authenticate clients using various mechanisms and to ensure traffic between the
+      server and the client is encrypted. This section provides a primer for the configuration
+      of listeners.</p>
+
+    <p>Kafka brokers support listening for connections on multiple ports. This is configured through
+      the <code>listeners</code> property in the server configuration, which accepts a comma-separated
+      list of the listeners to enable. At least one listener must be defined on each node. The format
+      of each listener defined in <code>listeners</code> is given below:</p>
```
Comment (jsancio) [inline docs/security.html]:
```suggestion
    <p>Kafka servers support listening for connections on multiple ports. This is configured through
      the <code>listeners</code> property in the server configuration, which accepts a comma-separated
      list of the listeners to enable. At least one listener must be defined on each server. The format
      of each listener defined in <code>listeners</code> is given below:</p>
```

File: docs/security.html [hunk 7]
```diff
@@ -36,7 +36,104 @@ <h3 class="anchor-heading"><a id="security_overview" class="anchor-link"></a><a
...
+    
+    <p>If each required listener uses a separate security protocol, it is also possible to use the
+      security protocol name as the listener name in <code>listeners</code>. Using the example above,
+      we could skip the definition of the <code>CLIENT</code> and <code>BROKER</code> listeners
+      using the following definition:</p>
+    
+    <pre class="line-numbers"><code class="language-text">listeners=SSL://localhost:9092,PLAINTEXT://localhost:9093</code></pre>
+      
+    <p>However, we recommend users to provide explicit names for the listeners since it
+      makes the intended usage clearer.</p>
```
Comment (jsancio) [inline docs/security.html]: I say we remove this if we don't want to recommend this type of configuration.

If a user is ready this section to understand an existing configuration maybe we can move this to the end of the section as a note.

File: docs/security.html [hunk 8]
```diff
@@ -36,7 +36,104 @@ <h3 class="anchor-heading"><a id="security_overview" class="anchor-link"></a><a
...
+
+    <p>Among the listeners in this list, it is possible to declare the listener to be used for
+      inter-broker communication by setting the <code>inter.broker.listener.name</code> configuration
+      to the name of the listener. If not defined, then the inter-broker listener is determined
+      by the security protocol defined by <code>security.inter.broker.protocol</code>, which
+      defaults to <code>PLAINTEXT</code>.</p>
```
Comment (jsancio) [inline docs/security.html]: Should we mentioned that inter-broker communication is mainly used for the replication of topic partitions?

File: docs/security.html [hunk 9]
```diff
@@ -36,7 +36,104 @@ <h3 class="anchor-heading"><a id="security_overview" class="anchor-link"></a><a
...
+      which makes it less likely for application traffic to impede timely propagation of metadata changes
+      (such as partition leader and ISR updates). Note that the default value is null, which
+      means that the controller will use the same listener as defined by <code>inter.broker.listener</code></p>
+    
+    <p>In a KRaft cluster, the listener defined by <code>inter.broker.listener.name</code> is used
+      exclusively for requests sent to nodes which have the "broker" role enabled in <code>process.roles</code>.
+      For legacy clusters, the active controller will connect to all brokers in the cluster
+      to send metadata updates. In KRaft, it is the other way around. The broker is responsible
+      for finding the current controller and connecting to it in order to receive metadata updates.
+      Hence the configuration <code>control.plane.listener.name</code> cannot be used in KRaft clusters.</p>
```
Comment (jsancio) [inline docs/security.html]:
```suggestion
    <p>In a KRaft cluster, the listener defined by <code>inter.broker.listener.name</code> is used
      exclusively for requests sent by brokers to brokers. A Kafka server is a broker if the <code>broker</code> role enabled in <code>process.roles</code>.
      In KRaft, the broker is responsible
      for finding the current controller and connecting to it in order to receive metadata updates.
      The broker determines the listener and security protocol based the configuration in <code>controller.listener.names</code> and <code>listener.security.protocol.map</code></p>
```

I think it is easier to read if we document how legacy cluster are configure in one paragraph/section and how kraft clusters are configured in another paragraph/section. I think that comparing the two makes it harder to read and understand.

I would argue that at this point the user it trying to configure their cluster correctly. They are less interest in the reason of why the configuration are different. I think they can read the KIPs and discussion for that kind of detail. What do you think?

File: docs/security.html [hunk 10]
```diff
@@ -36,7 +36,104 @@ <h3 class="anchor-heading"><a id="security_overview" class="anchor-link"></a><a
...
+      
+    <p>Instead, nodes with the "controller" role must define separate listeners through the
+      <code>controller.listener.names</code> configuration. Any node running with both the
+      "broker" and "controller" roles must use separate listeners for the broker and controller.
+      For example, a minimal configuration for a combined node might look like the following:</p>
```
Comment (jsancio) [inline docs/security.html]:
```suggestion
    <p>Instead, servers with the <code>controlle</code> role must define separate listeners through the
      <code>controller.listener.names</code> configuration. Any server running with both the
      <code>broker</code> and <code>controller</code> roles must use separate listeners for the broker and controller.
      For example, a minimal configuration for a combined server might look like the following:</p>
```

The broker also needs to the controller listener name so that it knows how which security protocol to use.

File: docs/security.html [hunk 11]
```diff
@@ -36,7 +36,104 @@ <h3 class="anchor-heading"><a id="security_overview" class="anchor-link"></a><a
...
+listener.security.protocol.map=BROKER:SASL_SSL,CONTROLLER:SASL_SSL</code></pre>
+
+    <p>This defines two listeners, one for the controller and one for the broker. It is conventional
+      to use a separate listener for client traffic as well.</p>
+    
+    <p>The controller will accept requests on all listeners defined by <code>controller.listener.names</code>.
+      For outbound requests to other controller nodes, the first listener listed in <code>controller.listener.names</code>
+      will be used. The list of listeners provides a way to change the active listener from one port
+      to another through a roll of the cluster. Basically do one roll to expose the new listener,
+      and one roll to remove the old listener.</p>
```
Comment (jsancio) [inline docs/security.html]:
```suggestion
    <p>The controller will accept requests on all listeners defined by <code>controller.listener.names</code>.
      For outbound requests to other controller servers, the first listener listed in <code>controller.listener.names</code>
      will be used. The list of listeners provides a way to change the active listener from one port
      to another through a roll of the cluster. Basically do one roll to expose the new listener,
      and one roll to remove the old listener.</p>
```

Hmm. Regarding the last sentence, maybe we can add a section at the end describing the process of changing the controller listener. What do you think? I think that last sentence is hiding a lot of details.

File: docs/security.html [hunk 12]
```diff
@@ -36,7 +36,104 @@ <h3 class="anchor-heading"><a id="security_overview" class="anchor-link"></a><a
...
+    
+    <p>In the following section, this document covers how to enable SSL on a listener for encryption
+      as well as authentication. The subsequent section will then cover additional authentication
+      mechanism using SASL.</p>
```
Comment (jsancio) [inline docs/security.html]: Should we have these relative references "In the following section" and "The subsequent section"? Should we instead use the anchor id to provide a direct reference to them.

Feel free to ignore this comment if it is not possible with the current HTML documentation implementation.

**Review by jsancio — COMMENTED** [review]
Thanks for the improvement to the documentation.

File: docs/security.html (hunk 9 above)
Comment (hachikuji) [reply docs/security.html]: Let me try to make the comment more concise. I just wanted a brief explanation for why the control plane listener doesn't exist in KRaft clusters.

File: docs/security.html (hunk 7 above)
Comment (hachikuji) [reply docs/security.html]: Below we still have references to this approach, so I wanted to mention it. In a subsequent patch, I was going to clean up those sections as well, so maybe we can remove this then as well.

**Review by jsancio — APPROVED** [review]
LGTM. This documentation helps to clarify how to configure brokers and controllers in KRaft!

**Comment by hachikuji** [comment]
@showuon I'd appreciate one more look before we merge if you have time. Thanks!

File: docs/security.html [hunk 13]
```diff
@@ -36,7 +36,136 @@ <h3 class="anchor-heading"><a id="security_overview" class="anchor-link"></a><a
...
+      protocol to another through a roll of the cluster (one roll to expose the new listener,
+      and one roll to remove the old listener). When multiple controller listeners are defined,
+      the first one in the list will be used for outbound requests.</p>
+
+    <p>It is conventional in Kafka to use a separate listener for clients. This allows the
+      inter-cluster listeners to be isolated at the network level. In the case of the controller
+      listener in KRaft, the listener should be isolated since clients do not work with it
+      anyway. Clients are expected to connect to any other listener configured on a broker.
+      Any requests that are bound for the controller will be forwarded as described
+      <a href="#kraft_principal_forwarding">below</a></p>
```
Comment (showuon) [inline docs/security.html:162]: I guess this is in a separate PR, right?

**Review by showuon — APPROVED** [review]
LGTM!

**Comment by showuon** [comment]
> @showuon I'd appreciate one more look before we merge if you have time. Thanks!

Will merge it soon.

**Comment by showuon** [comment]
backport to 3.3

File: docs/security.html (hunk 13 above)
Comment (hachikuji) [reply docs/security.html:162]: I added the principal forwarding section in a previous patch.
